11_notify/willshen8/cmd/puppy-server/main_test.go

```diff
@@ -0,0 +1,163 @@
+)
+
+// Unit test for parseCmdArgs() with long flags
+func TestParseCmdArgs(t *testing.T) {
+	args = []string{"--data", "/dev/null", "--port", "1234", "--store", "map"}
+	returnedArgs, _ := parseCmdArgs(args)
+	assert.NotNil(t, returnedArgs.file)
+	assert.Equal(t, 1234, returnedArgs.port)
+	assert.Equal(t, "map", returnedArgs.storer)
+}
+
+func TestParseCmdArgsWithError(t *testing.T) {
+	args = []string{"--wrongflag", "/dev/null", "-p", "1234", "-s", "map"}
+	_, err := parseCmdArgs(args)
+	assert.Error(t, err)
+}
+
+// Unit test for parseCmdArgs() with short flags
+func TestShortFlag(t *testing.T) {
+	args = []string{"-d", "/dev/null", "-p", "1234", "-s", "map"}
+	returnedArgs, _ := parseCmdArgs(args)
+	assert.NotNil(t, returnedArgs.file)
+	assert.Equal(t, 1234, returnedArgs.port)
+	assert.Equal(t, "map", returnedArgs.storer)
+}
+
+// Unit test (positive test) for validatePortFlag()
+func TestParseCorrectPort(t *testing.T) {
+	// setup the test server with arguments
+	args = []string{"--data", "/dev/null", "--port", "1234"}
+	parsedArgs, _ := parseCmdArgs(args)
+	flagReturned, _ := validatePortFlag(parsedArgs)
+	var expectedFlag = "1234"
+	assert.Equal(t, expectedFlag, flagReturned)
+}
+
+// Unit test (negative test) for validatePortFlag()
+func TestParseWrongPort(t *testing.T) {
+	// setup the test server with arguments
+	args = []string{"--data", "/dev/null", "--port", "123456789"}
+	parsedArgs, _ := parseCmdArgs(args)
+	flagReturned, flagError := validatePortFlag(parsedArgs)
+	var expectedFlag = "0"
+	var expectedError = "invalid port number entered"
+	assert.Equal(t, expectedFlag, flagReturned)
+	assert.Equal(t, expectedError, flagError.Error())
+}
+
+// Unit tests for createStorer()
+func TestCreateStorerWithCorrectArguments(t *testing.T) {
+	tests := []struct {
+		testName string
+		storer   string
+		err      error
+	}{
+		{testName: "map storer", storer: "map", err: nil},
+		{testName: "sync storer", storer: "sync", err: nil},
+	}
+	for _, tc := range tests {
+		tc := tc
+		t.Run(tc.testName, func(t *testing.T) {
+			_, actualErr := createStorer(tc.storer)
+			assert.Equal(t, tc.err, actualErr)
+		})
+	}
+}
+
+func TestCreateStorerWithWrongArguments(t *testing.T) {
+	tests := []struct {
+		testName string
+		storer   string
+	}{
+		{testName: "wrong storer", storer: "blah"},
+		{testName: "empty storer", storer: ""},
+	}
+	for _, tc := range tests {
+		tc := tc
+		t.Run(tc.testName, func(t *testing.T) {
+			_, actualErr := createStorer(tc.storer)
+			assert.Error(t, actualErr)
+		})
+	}
+}
+
+// Unit test for createPuppies()
+func TestSavePuppiesToStore(t *testing.T) {
+	var buf bytes.Buffer
+	out = &buf
+	var puppies []puppy.Puppy
+	puppies = append(puppies, puppy.Puppy{ID: 1, Breed: "Jack Russell Terrier", Color: "White and Brown", Value: "1500"})
+	storer := puppy.NewMapStore()
+	returnedPuppyStorer, _ := createPuppies(storer, puppies)
+
+	expected := "&{map[1:{1 Jack Russell Terrier White and Brown 1500}] 2 {0 0}}\n"
```

Comment:
Making a string from these store and comparing that is not really the best way to do this test:

1) You are vulnerable to the internal details of the storer affecting the test, as seen by the change you needed to make for this test to pass, just because you added a mutex to the storer,
2) It may be wrong. What if, for example, the dog breed stored was "Jack Russell" and the colour was "Terrier White and Brown" - it looks identical but is different. This is not likely to happen in this specific case but it does give an example of how this sort of testing can go wrong.

Better would be to compare the contents of two stores directly, or to test that the resulting store contains what you are expecting it to contain. Consider a storer that is implemented with an external SQL database - you cannot just compare the store - all you can do is check it contains what you expect it to. It may contain other things too, but your storer_test can do deeper tests that the storer does the right thing in that regard.

The best way to think about testing a function (in this case, `createPuppies`) is to look at what it does and nothing else. In this case, given a slice of puppies, it puts them in a store. So the validation is that the puppies in the slice are in the store. If there is something else in the store, that is irrelevant - you are not testing that the store works properly, just that `createPuppies` does what it claims.

No need to change this though - it's from a previous lab - I just wanted to comment and give some feedback.

Reply:
Thanks very much for this great input. I will take care of it in the future. If I would do it again, I would test this way (based on your feedbacks) - there is no error when creating puppies and puppies are stored successfully. Agree that using these internal string representations of store is not ideal.